Add torch.profile benchmarking function to feedback_fns

exclamaforte opened the pull request:
Summary: Updates some benchmarking code to have the option to use torch.profile, and passes in a thunk to benchmark_fns to get this information (this will be a different result from `timings`, which are already passed into those functions).

Test Plan: Existing unit tests.

Differential Revision: D74444990




cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @chenyang78 @kadeng @muchulee8 @amjames @chauhang @aakhundov

eellison commented:
and @nmacchioni who has been wary of adding complexity to this file.

exclamaforte commented on torch/_inductor/select_algorithm.py:
```diff
@@ -2236,8 +2256,26 @@ def do_autotuning(choices, precompile_fn):
+                # Avoid benchmarking in a separate process because it's not easy to signal to the TuningProcess that we
+                # should use the profiler.
+                with config.patch(
+                    profile_bandwidth_with_do_bench_using_profiling=True,
```
So `profile_bandwidth_with_do_bench_using_profiling` basically tells the `CachingAutotuner` to use the profiler. The `AlgorithmSelectorCache` doesn't use the `CachingAutotuner`, it calls the benchmark method of `ChoiceCallers`, which is why it's necessary to change this in many spots. The do_bench function is InductorBenchmarker/TritonBenchmarker.bench in this case, so we're replacing that with do_bench_using_profiling.

> For autotune_in_subproc=False, it seems a bit weird to me to quietly not use the subproc-based method if the user requested it.

I can support autotune_in_subproc, but I would need to signal to the `TuningProcess` that it should use the benchmarker somehow, because the subprocs are all created at startup so just setting the envvar doesn't work. In status quo, they're all doing profiler benchmarking or none of them are. I can make this refactor but I don't think it's worth complicating `TuningProcess`. I'd rather just run them in the single process, since this change is mainly focused on data collection/logging.

masnesral replied:
> I can support autotune_in_subproc

Right. I wasn't proposing to support it. Isn't the new feature you're adding optional, or did I misread and it's active by default? I was just saying that we could fail if user wants both profiling and subprocs rather than quietly doing something else.

masnesral replied:
Does that mean you can remove `autotune_in_subproc=False,` here?

exclamaforte replied:
no, because for this logging code, we need it to switch between profiler and not-profiler, which only works if it's not using the subprocs

exclamaforte committed:
Add torch.profile benchmarking function to feedback_fns (#153579)

Summary:
Updates some benchmarking code to have the option to use torch.profile, and passes in a thunk to benchmark_fns to get this information (this will be a different result from `timings`, which are already passed into those functions).

Pull Request resolved: https://github.com/pytorch/pytorch/pull/153579

Test Plan:
Existing unit tests.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

Reviewed By: masnesral, coconutruben

Differential Revision: D74444990

Pulled By: exclamaforte

exclamaforte commented:
@nmacchioni Here's the full stack of internal changes: https://www.internalfb.com/diff/D75489476

nmacchioni commented:
> There is a desire to extend this data collection to isolate the kernel runtimes from the cpu overhead, which is the use case for do_bench_using_profiling.

My understanding is that do_bench already differentiates between CPU overhead and GPU execution duration, have you had an issue with do_bench?

nmacchioni commented:
> > I will be ripping out the subprocess autotuning soon
> 
> @nmacchioni Oh wait, don't do that yet. Pieces of that may still be useful for remote autotuning.

Let's chat offline
